File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if head == None or head.next ==None:
            return 
        
        slow = head
        fast = head

        while fast !=None and fast.next!=None:
            slow = slow.next
            fast = fast.next.next

        curr = slow.next
        prev = None

        slow.next = None

        while curr!=None:
            nxt = curr.next

            curr.next = prev

            prev = curr
            curr = nxt
        
        first = head 
        second = prev

        while second!=None:
            firstNext = first.next
            secondNext = second.next

            first.next = second
            second.next = firstNext

            first=firstNext
            second= secondNext

File: 0143-reorder-list/0143-reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {

        if (head == nullptr || head->next == nullptr)
            return;
        ListNode* slow = head;
        ListNode* fast = head;

        while(fast!=nullptr && fast->next!=nullptr){
            slow = slow->next;
            fast = fast->next->next;
        }

        //everse after  slow 

        ListNode* prev = nullptr;
        ListNode* curr = slow->next;
        // disconnect 2nd half
        slow->next = nullptr;

        while(curr!=nullptr){
            ListNode* nxt = curr ->next;

            curr->next = prev;

            prev = curr;
            curr = nxt;
        }

        //merge 2 halfes alternativvely 
        ListNode* first = head;
        ListNode* second = prev ;

        while(second!=nullptr){
            ListNode* firstNext = first->next;
            ListNode* secondNext = second->next;

            first->next = second;
            second->next = firstNext;

            first = firstNext;
            second = secondNext;
        }
        
    }
};